### PR TITLE
add plotting examples for geospatial examples

### PR DESCRIPTION
**Pull request by ljwolf**
This adds a few examples of geospatial networks for points, lines, and polygons. A few questions. 

1. It makes a bit more sense to show how you build a graph built from a given type of geometric data (like, points, lines, or polygons) rather than from a specific package. I hope this makes it more clear: `geopandas` for reading, `pysal` for building the topology. @martinfleis, you may want to jump in here to demonstrate `momepy.gdf_to_nx` for LineString-based geometries? The queen graph on raw segments, while correct, is not as abstract as what folks generally want. 
2. In the `plot_polygons.py` example, I opted for an online geojson dataset, rather than a built-in dataset. If that's alright, then we can proceed. I'd recommend showing that it's not quite important whether it's a shapefile, geojson, geopackage, etc. If `geopandas` can read it, it shouldn't matter. 
3. Should I just commit the required files to this examples directory here? Or, is it sufficient to show that we are just reading in from a specific path?

### CODE REVIEW

**Comment by jarrodmillman** [comment]
Thanks!  This is a huge help.

To fix the failing test add ``contextily`` to ``requirements/example.txt``.

It might be nice to have at least one example of reading a local file.  However, the online geojson dataset probably suffices (assuming the dataset is not likely to move or disappear).  @dschult @rossbar @MridulS what do you think about whether we need to have the files stored in the directory or not?

It would be nice if you could beef up the docstring to say a little more about the various packages being used in each example and perhaps pointing folks to the project home page, documentation, and/or galleries as appropriate.

**Comment by MridulS** [comment]
Thanks @ljwolf for this!

> what do you think about whether we need to have the files stored in the directory or not?

If the dataset (or a slice of it) is small enough <0.5MB, keeping it inside the example repo should be perfectly fine IMO.

File: examples/geospatial/plot_lines.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+# read in example data online
+filepath = examples.get_path("streets.shp")
+tempe = geopandas.read_file(filepath).to_crs(epsg=3857)
+
+coordinates = numpy.column_stack((tempe.centroid.x, tempe.centroid.y))
+
+# construct the graph
+w_graph = weights.Queen.from_dataframe(tempe)
+graph = w_graph.to_networkx()
```
Comment (ljwolf) [inline examples/geospatial/plot_lines.py:26]: This should move down below the next heading, and bump `positions` down further still.

**Comment by dschult** [comment]
Thanks for this!

The geojson file in question is ~200KB. The URL is likely to cause more trouble simply because we don't know when it will change. If we put the file into our repo then we know when it will change.  That's my reason for wanting it in the repo rather than via https.  I may be missing something here though. I'm guessing the data won't need to be updated often -- or at least if it does, the URL would need to change anyway, so we're not adding to our work to keep in in the repo.

Nice work! Thanks!

**Comment by martinfleis** [comment]
Talking about lines, there are generally two ways of creating graph object from geometry representing planar (street) network.

A) a **primal** approach, where each intersection is a node and each linestring segment connecting two intersections is an edge. That is the model used for 90% of network analysis in urban science to my knowledge. That is how OSMnx builds its `networkx.MultiDiGraph` and that is one of the options momepy's `gdf_to_nx` can do, because that is what people usually expect.

B) a **dual** approach, where each line is a node and intersection topology is turned into edges. That is in the example @ljwolf wrote. One of the options how this is used for street network analysis is an angular analysis, where your routing is weighted via angles between street segments on intersections. The idea is that for better flow, people prefer fewer turns over a shorter distance. But to actually use it you need to get the angle attribute on the edges.

Both of these can be done using `momepy.gdf_to_nx` as Levi mentioned above. Plus you can rebuild geometry again with `momepy.nx_to_gdf` after your graph analysis. 

What I would propose is to add a bit of description to docstring to clarify the use case of Queen contiguity as people may eventually expect it is a primal graph although it is not. Also, if you networkx folks are interested, I can draft another example using momepy to generate both options and go back to GeoDataFrames to have the use cases covered more comprehensively.

**Comment by ljwolf** [comment]
I think documenting both are important, but I would definitely prefer to just suggest folks use `momepy` for this! Document the "right way" once, clearly, using a tool built for the job. 

Just blow away the code in `plot_lines.py` and replace it with the correct momepy implementation!

**Comment by jarrodmillman** [comment]
@martinfleis We recently deprecated ``nx_shp`` https://github.com/networkx/networkx/pull/4298.  As you probably know, we (nx core developers) don't have much familiarity with geospatial data.  Our CI tests don't even include an example shapefile, so I suspect ``nx_shp`` is rarely (or ever) the recommended way to read/write shapefiles.  A quick search on stackoverflow makes that concern seem reasonable.  We also kept getting PRs fixing and issues mentioning issues with ``nx_shp``, which we were unable to confidently respond to.

Since there are several established geospatial data analysis packages for Python, our goal is to provide simple examples in our gallery for how to use these packages with NetworkX.  Ideally, these examples will provide examples of the "right way" to do things as @ljwolf suggested.  @gboeing also agreed to provide an example of working with OSMnx and NetworkX.

Once we have the geospatial gallery examples in place, I hope that users will end up being steered to the right geospatial libraries.  This will also allow us to make sure we work well with pysal, geopandas, momepy, and osmnx, since we will be running them now as part of our CI system.

Since these examples are meant to guide users to the appropriate libraries and to point users to the recommend way to handle work with geospatial data, it would be great if these examples could include more explanatory text than some of our existing examples.  In particular, it would be great if the examples could provide a brief description of the various packages being used in each example and pointing folks to the project home page, documentation, and/or galleries as appropriate.

We use sphinx-gallery for our example gallery:
- https://networkx.org/documentation/latest/auto_examples/index.html

You can read more about the syntax used to provide a more notebook-like structure here:
- https://sphinx-gallery.github.io/stable/tutorials/plot_parse.html#sphx-glr-tutorials-plot-parse-py

It would be great if @jorisvandenbossche could also take a look at these examples.  Since we (the NX core developers) don't really know this field, we will have to rely on you folks to provide the examples and to ensure we are leading our users astray.

This is a priority for us, so we are more than happy to quickly review, provide feedback, and answer questions.

**Comment by jarrodmillman** [comment]
Also, if you need any additional packages installed to run the examples, just add them to ``requirements/example.txt``.

**Comment by jarrodmillman** [comment]
We use black for formatting.  So you should be able to fix the style and format test failure by installing black
```
pip install black
```
and running it on the files you are changing.  You can automate this using a pre-commit hook
```
pip install -r requirements/developer.txt
pre-commit install
```

**Comment by jarrodmillman** [comment]
LGTM!  @dschult , @rossbar, @MridulS  may have additional comments and suggestions.  I am also hoping that @martinfleis and @gboeing will have time to take a quick look and provide feedback.  Let us know if you have any more questions or suggestions.

I started adding a short section on reading and writing graphs to our tutorial (#4283).  I would like to add a few sentences about geospatial data and point users to the main packages for geospatial data analysis in Python.  If you have time and / or interest, feel free to suggest text for that PR.  In particular, would it be reasonable to point people to:
 https://geographicdata.science/book/intro.html

We have a few users that have been asking for better support for geospatial data in NetworkX.  I am going to point them to this PR.

**Comment by jarrodmillman** [comment]
@brendan-ward in https://github.com/geopandas/geopandas/issues/1592 you suggested it may make sense to provide "a series of examples on how to use GeoPandas + NetworkX" together.  I am not sure if all the examples belong here (or even how many examples are needed) since I don't work with geospatial data.  So I am hoping you could take a look at these examples and provide feedback.  Also, if you have additional examples that you would like to see included please let us know or better yet (since none of the NetworkX developers are in a position to provide additional examples) just create a new PR including those examples.

**Comment by NicolasCadieux** [comment]
Hi,

If you don't want to be flooded with questions, examples should support or detect parallel edges, 3D geometries and line collections (multilinestrings). We should also have examples to show how to add disconnected nodes (like a street adresse) to the network. Having shapefiles or a geopackage to the example could be nice but examples could also be simple WKT examples in a text file.

I could help out here.

**Comment by martinfleis** [comment]
I will prepare an example on street networks using momepy (as outlined above). That can handle parallel edges (it is a `MultiGraph`). OSMnx can handle then as well (`MultiDiGraph`). Momepy also supports 3D geometry.

@NicolasCadieux Regarding MultiLineStrings, how do you envisage it should work? The only idea coming to my mind is to explode it before graph creation but it is probably not what you have in mind?

**Comment by NicolasCadieux** [comment]
Hi,
I haven't played with Momepy. I have to check that out. If it used shapely, then some manipulations with not work with 3D objects and the shapely error is confusing.  MultiLineStrings could be exploded if the are connected Linestrings with no flip in directions.  If they have flips, of if they are disconnected, then we should flag the error as the used probably does not know he has them.  Using as graph that can handle parallel edges is good but much slower and tough on memory.  I would definitely use one in one example but I would also put a digraph example and flag the parallel edges as they show up, or modify them to eliminate beginners errors.  This way, we show them how to handle this situation.

**Comment by NicolasCadieux** [comment]
We could used or refer to this https://www.mdpi.com/2306-5729/5/1/8 in text if you don't want to start from scratch.

**Comment by NicolasCadieux** [comment]
If we choose to explode the MultiLineStrings, we need to keep in mind that any weight value would need to be proportional to the line segments.  That could be ok for stuff like distances but not stuff like distances from radiation source or some weight that would decrease with an inverse square.  It might just be better to flag them out. 

In my article, I explain some of the thing that need to be done to a GIS network for it to work on a graph. It would be nice to add some of this info in the GIS examples.

**Comment by NicolasCadieux** [comment]
I see momepy is built on top of geopandas so we would need to be careful with 3D shapes.  Some shapely manipulation will not work. The old nx shape read had this problem so you could test with that to see the error that is given.

**Comment by martinfleis** [comment]
> I see momepy is built on top of geopandas so we would need to be careful with 3D shapes. Some shapely manipulation will not work. The old nx shape read had this problem so you could test with that to see the error that is given.

What I meant is that GeoDataFrame of LineStringZ geom type (x, y, z coords) gets converted to `nx.MultiGraph` and back without the loss of 3D data and the creation reflects all three dimensions. Since its main aim is to work with street networks, I haven't tried much than that.

**Comment by NicolasCadieux** [comment]
Ok,

Start with what you are proposing and then I can test to see if it bugs. It's probably all good if the shapely objects found in edges are not modified or manipulated. I will check it out. Tell me when and if you want my input.

Nicolas

**Comment by brendan-ward** [comment]
Big thanks to @ljwolf for adding examples here!

Depending on file size, it would be good to use one format for example files; I'd recommend against using shapefiles here since that clutters up the directory.  GeoJSON - while verbose - may be a good option if not too big here, since it will also work with Github's builtin GeoJSON viewer.


A few suggestions (toward the broader intent of these examples, not all need to be in this PR), that generally follow in line with comments raised above:

It would be very good to add a bit to the `examples/geospatial/readme.txt` (could that be converted to a *.md for better formatting?) a bit of additional context that will help folks understand the examples a bit better.   In addition to crosslinks to each of the packages used here (GeoPandas, PySAL, momepy), it would be good to briefly give a little bit more info about what problem they solve in general, and also provide a bit more info about some of the abstractions here.

I'm thinking something like:
```
The following geospatial examples showcase different ways of performing network analyses using packages within the geospatial Python ecosystem.  Example spatial files are stored directly in this directory.

## Dependencies

[GeoPandas](https://geopandas.readthedocs.io/) provides interoperability between geospatial formats and storage mechanisms (e.g., databases) and Pandas data frames for tabular-oriented processing of spatial data, as well as a wide array of supporting functionality including spatial indices, spatial predicates (e.g., test if geometries intersect each other), spatial operations (e.g., the area of overlap between intersecting polygons), and more.

[PySAL](https://pysal.org/) provides a rich suite of spatial analysis algorithms.  From a network analysis context, [spatial weights](https://pysal.org/libpysal/api.html#spatial-weights) provide... (Levi please add more here).

See the following examples that use PySAL:
* `examples/geospatial/plot_delaunay.py`
* ... (depends on split between PySAL and Momepy examples)

[momepy](http://docs.momepy.org/en/stable/) builds on top of GeoPandas and PySAL to provide a suite of algorithms focused on urban morphology.  From a network analysis context, momepy enables you to... (Martin please add more here).

See the following examples that use momepy:
* ...

## Key Concepts
One of the essential tasks in network analysis of geospatial data is defining the spatial relationships between spatial features (points, lines, or polygons).

`PySAL` provides several ways of representing these spatial relationships between features using the concept of spatial weights.  These include relationships such as `Queen`, `Rook`, ...(Levi please add more here with a brief explanation of each).

(Martin please add a brief intro to abstractions used in your examples)
```

@ljwolf @martinfleis descriptions of your packages above for example purposes only, apologies for any incorrect statements, etc!

I think one of the things I struggled with a bit in the examples and discussion in this PR is that these libraries provide very useful abstractions from a network analysis standpoint - but they abstractions, and also a require a certain degree of domain knowledge or context in order to understand what problem them solve (as demonstrated by the examples).  I think in particular that briefly defining some of the core concepts that are not self-evident from the example code (e.g., `Queen` contiguity), with pointers over to more documentation in each of these libraries would help alleviate that.

I think when you get into more advanced terms / concepts (like "primal" above), it will be particularly helpful to briefly describe them (and crosslink to more details).  This may better enable folks to see how to adapt these examples to different domains where those terms are not used.

I'd like to see an additional example that doesn't rely on PySAL or momepy, and is focused less on the network analysis outcome and more on solving a geospatial problem, that happens to leverage characteristics of the underlying graph.  To be clear, I'm sure these problems could be solved better and with less code by leveraging PySAL - but perhaps showing a more direct example with fewer abstractions may help show what is possible - and the value in deferring some of that to PySAL 😄 .

The one that comes to mind from my work is a simple "dissolve" of adjacent polygons.  I'll try to work up an example based on `nuts1.geojson` here.

**Comment by ljwolf** [comment]
> I'd like to see an additional example that doesn't rely on PySAL or momepy, and is focused less on the network analysis outcome and more on solving a geospatial problem, that happens to leverage characteristics of the underlying graph.

To build the adjacency matrix from them, `pygeos` should be plenty fast for all-pairs touches among polygons using the `STRTree`, but that's a bit more complex than I thought appropriate at first brush..... I'm happy to add an all-pairs-touches with `pygeos` to the `polygon` example and explain the `Queen`/`Rook` jargon. 

However, I would *not* want to document this as an all-pairs `shapely` operation, as that scales very badly, and I think we should recognize folks take examples as encouragements of a specific analytical pattern. 

A dissolve of the connected components would be real neat and clear w/ the NUTS1 regions. This kind of "stringing packages together" was a serious motivation for [geographicdata.science/book](https://geographicdata.science/book) as well as our book series colleagues at [pythongis.org](https://pythongis.org/), since it quickly gets very long to do it from the basic libraries or touches on many packages to do it "right" at scale. 

But, I really think examples should be highly germane to `networkx`... so would this GIS-style dissolve be better in `geopandas` or a dedicated `geograph` package we may yet write to centralize all of these toplogy-of-geographic data concerns?

**Comment by jarrodmillman** [comment]
@brendan-ward Thanks for the comments.  I have two **minor** points:

1.  I like the proposed text / direction for the readme.  However, the ``examples/geospatial/readme.txt`` should be written with reStructuredText (sphinx-gallery doesn't support markdown by default and I can't find an option to change that).  I also may end up moving some of that text to the tutorial (or somewhere else) b/c it doesn't look great on the gallery page.  But that can be in a future cleanup PR that we (the core NX developers) can handle.  So just focus on the text for now.  (For that matter, we can even change the format from markdown to reStructuredText later.)
2.  I prefer having different format files, since it will demonstrate that different file formats can be used.  However, maybe it suffices to just add a comment above every file reading bit stating that many different file formats can be used (and maybe list a few).  In fact, it is probably worth adding the comment about the different file formats above each file read/write line regardless of whether the examples all use the same format or not.

**Comment by gboeing** [comment]
FYI @jarrodmillman (and others), following up on our email convo, I've been wrapping up semester teaching and grade, and should finally have some time this week to open a draft PR with an OSMnx example.

**Comment by brendan-ward** [comment]
@ljwolf 
> I'm happy to add an all-pairs-touches with pygeos to the polygon example

Would this be simpler, more performant, or more of a recommended pattern than the example I added in #4381?  If so, we can certainly drop that PR in light of changes here.

For that example, I was trying to leverage geopandas builtins (sjoin, dissolve) to keep that somewhat high level; it could be adapted to just use pygeos and pandas directly, but would be a bit more verbose and potentially harder to follow.  Under the hood, it uses the spatial index to help keep the all-pairs-touches fast (note: it leverages intersection predicate built into STRtree when that backend is used).  I completely agree about wanting to avoid patterns that scale poorly.

I don't have strong opinions about where that example lives.

@jarrodmillman 
> just add a comment above every file reading bit stating that many different file formats can be used

I think a comment where `geopandas.read_file` or top of the file should suffice (e.g., 'You can use any geospatial format or data source supported by your installation of GeoPandas, including GeoJSON, ESRI Shapefile, Geopackage, etc.'); it seems appropriate to defer examples of support for different file formats to GeoPandas and focus the examples here more on what you can then do with the GeoDataFrames that you get from those.

**Comment by jarrodmillman** [comment]
@ljwolf I suspect you plan to do more work on this before marking it as ready to review.  But I wonder if we should merge this now (with https://github.com/ljwolf/networkx/pull/1) and then you can make a new PR to fix things up.

The reason being that others may be making PRs that could benefit from having the data files and updtaed ``requirements/example.txt`` included.  We are already running into the missing data file issue with #4381.  Alternatively, we could make a new PR with just the data files,  ``requirements/example.txt``, and maybe the updated ``README.txt``.  Then you could rebase this PR on master...

Either option is fine with me or we can do something else.  Thoughts?

**Comment by iboates** [comment]
I have been away from this from some time as I was focusing on finishing some personal projects but I see that I have a few mentions. I actually haven't heard of pySAL before but it looks pretty solid, and I see that it is maintained in parallel with Geopandas, so I would trust that it would be the recommended approach. In fact I may go back and look at some of my own code to see if I can replace some of my own hacky implementations with some nice optimized one-liners from PySAL.

The truth is I don't think I'm really an expert on hardcore code optimization so I think that whatever PySAL has implemented is probably going to be more performant than what I came up with already. I'm also just happy that there is a tighter integration between Geopandas and NetworkX because I use the two quite frequently and it's nice to be getting some "official" ways to work with the two in tandem.

File: examples/geospatial/plot_delaunay.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+"""
+=====
+Graphs from geographic points
```
Comment (rossbar) [inline examples/geospatial/plot_delaunay.py:3]: There are two examples [in the gallery](https://807-890377-gh.circle-artifacts.com/0/doc/build/html/auto_examples/index.html#geospatial) with the same title. It seems like this example could be named differently, maybe "Delaunay graph from geometric points" or similar?

File: examples/geospatial/plot_delaunay.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+from libpysal import weights, examples
+from libpysal.cg import voronoi_frames
+from contextily import add_basemap
+import matplotlib.pyplot as plt
+import networkx as nx
+import numpy
+import geopandas
+
+# read in example data online
```
Comment (rossbar) [inline examples/geospatial/plot_delaunay.py:21]:
```suggestion
# read in example data in GeoPackage format
```
or similar

File: examples/geospatial/plot_lines.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+"""
+
+from libpysal import weights, examples
+from contextily import add_basemap
+import matplotlib.pyplot as plt
+import networkx as nx
+import geopandas
+import numpy
+
+# read in example data online
```
Comment (rossbar) [inline examples/geospatial/plot_lines.py:18]:
```suggestion
# read in example data from a shapefile
```

File: examples/geospatial/plot_points.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+"""
+
+from libpysal import weights, examples
+from contextily import add_basemap
+import matplotlib.pyplot as plt
+import networkx as nx
+import numpy
+import geopandas
+
+# read in example data online
```
Comment (rossbar) [inline examples/geospatial/plot_points.py:20]: Similar to above comments

File: examples/geospatial/plot_polygons.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+as well as other kinds of graphs from the polygon centroids. 
+"""
+
+from libpysal import weights
+import matplotlib.pyplot as plt
+import networkx as nx
+import geopandas
+import numpy
+
+# read in example data online
```
Comment (rossbar) [inline examples/geospatial/plot_polygons.py:18]:
```suggestion
# read in example data in GeoJSON format
```

**Review by rossbar — APPROVED** [review]
Thanks very much for this @ljwolf and reviewers! I think this is a big step in the right direction. I agree with @jarrodmillman that it would be good to get this in as it serves as a nice nucleus (along with the osmnx example) for working with geospatial data. 

I took the liberty of making a few small changes (numpy import convention and sphinx headings) and have a few comments, namely:
 - The comments say "load data from online" but the data is local now. I left suggestions but feel free to reword as you see fit.
 - Two of the examples have the same name in the gallery, so it would be good to rename one of them.

Assuming there are no major sticking points for the other reviewers with geospatial domain expertise, I'm +1 on this going in and tackling further comments/text/improvements in follow-up PRs.

**Comment by rossbar** [comment]
A rebase on master should fix the circleCI build error

**Comment by martinfleis** [comment]
@ljwolf I've just made a PR (#4407) effectively replacing your `plot_lines.py` as you suggested above. I have covered both momepy and PySAL options there, so we can drop it from this PR.

**Comment by ljwolf** [comment]
Sounds great! I'll clean this up shortly with the recommended edits and then we can close these! Thanks Martin!

File: requirements/example.txt [hunk 7]
```diff
@@ -7,5 +7,7 @@ Cartopy>=0.18.0
 geopandas>=0.8
 osmnx>=0.16
 pysal>=2.3
```
Comment (martinfleis) [inline requirements/example.txt:9]:
```suggestion
libpysal>=2.3
```

**Comment by ljwolf** [comment]
@knaaptime, it might also be great to get a boundary length-proportional example in this form as well... I'm not sure where that code @renanxcortes & I cooked up lives now (`pysal/segregation`?).... but it should move to `libpysal` & get documented somewhere like here. Will follow up in the development call this month.

File: examples/geospatial/plot_delaunay.py (hunk 2 above)
Comment (ljwolf) [reply examples/geospatial/plot_delaunay.py]: Yes, I've made this more specific in the new text.

File: examples/geospatial/plot_delaunay.py (hunk 3 above)
Comment (ljwolf) [reply examples/geospatial/plot_delaunay.py]: I've added a little more explanation to each format's read.

**Comment by ljwolf** [comment]
The issue in the build seems to be with `pygeos` installation from pypi. I only install from conda, so I'm not quite sure how to get the right C version here. I am OK with removing this example if that complicates the stack...

**Comment by jarrodmillman** [comment]
You probably saw this, but we are getting this error now:
```
Unexpected failing examples:
/home/circleci/repo/examples/geospatial/plot_polygons.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/circleci/repo/examples/geospatial/plot_polygons.py", line 64, in <module>
    pygeos_polygons = pygeos.from_shapely(european_regions.geometry)
  File "/home/circleci/repo/venv/lib/python3.8/site-packages/pygeos/io.py", line 208, in from_shapely
    raise ImportError(
ImportError: The shapely GEOS version (3.7.1-CAPI-1.11.1 27a5e771) is incompatible with the GEOS version PyGEOS was compiled with (3.8.1-CAPI-1.13.3)
```

**Comment by martinfleis** [comment]
> The issue in the build seems to be with pygeos installation from pypi. I only install from conda, so I'm not quite sure how to get the right C version here. I am OK with removing this example if that complicates the stack...

You can do the same principle using `gdf.sindex.qury_bulk` which will use `rtree` in this case instead of `pygeos`. So staying within geopandas all the time should resolve it.

**Comment by ljwolf** [comment]
I'd rather just remove the example and ship as is, with the acknowledgement at the tail that this _can_ be done in `pygeos`, and I'll shoot that to `pygeos`.

**Pull request merged**